tools/nnyeah/nnyeah/Reworker.cs

```diff
@@ -0,0 +1,488 @@
+using System;
+using System.IO;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+using Mono.Cecil.Rocks;
+using System.Linq;
+using System.Collections.Generic;
+
+#nullable enable
+
+namespace nnyeah {
+	public class Reworker {
+		Stream stm;
+		ModuleDefinition? module;
+		TypeDefinition? embeddedAttributeTypeDef;
+		TypeDefinition? nativeIntegerAttributeTypeDef;
+		TypeReference? nativeIntegerAttributeTypeRef;
+		TypeReference? compilerGeneratedAttributeTypeRef;
+		TypeReference? embeddedAttributeTypeRef;
+		TypeReference? nintTypeReference;
+		TypeReference? nuintTypeReference;
+
+		Dictionary<string, Transformation>? methodSubs;
+		Dictionary<string, Transformation>? fieldSubs;
+
+		public Reworker (Stream stm)
+		{
+			this.stm = stm;
+		}
+
+		public void Load ()
+		{
+			if (module is not null)
+				return;
+			module = ModuleDefinition.ReadModule (stm);
+		}
+
+		public bool NeedsReworking ()
+		{
+			CheckModule ();
+
+			// simple predicate for seeing if there any references
+			// to types we need to care about.
+			// IntPtr is for future handling of types that
+			// descend from NObject and need to have the constructor
+			// changed to NHandle
+			return module!.GetTypeReferences ().Any (
```

Comment:
The compiler here is complaining because it is missing information, and that is why you need the !. The way you are doing it, it will required you to use ! in all calls that are done after CheckModule which is very very annoying. 

Rather than requiring to use the ! operator in every method that has called CheckModule, you can do the following:
```csharp
static void CheckModule ([NotNull] ModuleDefinition? module) {
    // you throw here
}
```
Then use CheckModule(module). That will allow you to avoid the ! operator. But maybe NeedsReworking could be an extension method on ModuleDefinition which then will only need to check null in the extension method once.